File: synapse-qubit-flow/src/parser/lexer.ts
export enum TokenType {
  // Keywords
  CIRCUIT = 'CIRCUIT',
  VQE = 'VQE',
  QAOA = 'QAOA',
  BACKEND = 'BACKEND',
  MEASURE = 'MEASURE',
  INTO = 'INTO',
  GRAPH = 'GRAPH',
  NODES = 'NODES',
  EDGES = 'EDGES',
  ANSATZ = 'ANSATZ',
  HAMILTONIAN = 'HAMILTONIAN',
  OPTIMIZER = 'OPTIMIZER',
  PROBLEM = 'PROBLEM',
  LAYERS = 'LAYERS',
  MIXER = 'MIXER',
  PROVIDER = 'PROVIDER',
  DEVICE = 'DEVICE',
  SHOTS = 'SHOTS',
  OPTIMIZATION = 'OPTIMIZATION',

  // Types
  ANGLE = 'ANGLE',
  INTEGER_TYPE = 'INTEGER_TYPE',
  REAL_TYPE = 'REAL_TYPE',

  // Gates
  H = 'H',
  X = 'X',
  Y = 'Y',
  Z = 'Z',
  RX = 'RX',
  RY = 'RY',
  RZ = 'RZ',
  CNOT = 'CNOT',
  CZ = 'CZ',
  SWAP = 'SWAP',
  T = 'T',
  S = 'S',

  // Pauli operators
  I = 'I',

  // Optimizer types
  COBYLA = 'COBYLA',
  SPSA = 'SPSA',
  ADAM = 'ADAM',

  // Provider types
  IBM = 'IBM',
  IONQ = 'IONQ',
  AWS = 'AWS',
  SIMULATOR = 'SIMULATOR',

  // Literals
  INTEGER = 'INTEGER',
  REAL = 'REAL',
  STRING = 'STRING',
  BOOLEAN = 'BOOLEAN',
  IDENTIFIER = 'IDENTIFIER',

  // Operators
  PLUS = 'PLUS',
  MINUS = 'MINUS',
  MULTIPLY = 'MULTIPLY',
  DIVIDE = 'DIVIDE',
  ASSIGN = 'ASSIGN',

  // Delimiters
  LPAREN = 'LPAREN',
  RPAREN = 'RPAREN',
  LBRACE = 'LBRACE',
  RBRACE = 'RBRACE',
  LBRACKET = 'LBRACKET',
  RBRACKET = 'RBRACKET',
  COMMA = 'COMMA',
  COLON = 'COLON',
  SEMICOLON = 'SEMICOLON',

  // Special
  PI = 'PI',
  THETA = 'THETA',
  EOF = 'EOF',
  NEWLINE = 'NEWLINE',
  WHITESPACE = 'WHITESPACE',
  COMMENT = 'COMMENT'
}

export interface Token {
  type: TokenType;
  value: string;
  line: number;
  column: number;
}

export class Lexer {
  private input: string;
  private position: number = 0;
  private line: number = 1;
  private column: number = 1;

  private keywords: Map<string, TokenType> = new Map([
    ['circuit', TokenType.CIRCUIT],
    ['vqe', TokenType.VQE],
    ['qaoa', TokenType.QAOA],
    ['backend', TokenType.BACKEND],
    ['measure', TokenType.MEASURE],
    ['into', TokenType.INTO],
    ['graph', TokenType.GRAPH],
    ['nodes', TokenType.NODES],
    ['edges', TokenType.EDGES],
    ['ansatz', TokenType.ANSATZ],
    ['hamiltonian', TokenType.HAMILTONIAN],
    ['optimizer', TokenType.OPTIMIZER],
    ['problem', TokenType.PROBLEM],
    ['layers', TokenType.LAYERS],
    ['mixer', TokenType.MIXER],
    ['provider', TokenType.PROVIDER],
    ['device', TokenType.DEVICE],
    ['shots', TokenType.SHOTS],
    ['optimization', TokenType.OPTIMIZATION],
    ['angle', TokenType.ANGLE],
    ['integer', TokenType.INTEGER_TYPE],
    ['real', TokenType.REAL_TYPE],
    ['H', TokenType.H],
    ['X', TokenType.X],
    ['Y', TokenType.Y],
    ['Z', TokenType.Z],
    ['RX', TokenType.RX],
    ['RY', TokenType.RY],
    ['RZ', TokenType.RZ],
    ['CNOT', TokenType.CNOT],
    ['CZ', TokenType.CZ],
    ['SWAP', TokenType.SWAP],
    ['T', TokenType.T],
    ['S', TokenType.S],
    ['I', TokenType.I],
    ['COBYLA', TokenType.COBYLA],
    ['SPSA', TokenType.SPSA],
    ['ADAM', TokenType.ADAM],
    ['IBM', TokenType.IBM],
    ['IonQ', TokenType.IONQ],
    ['AWS', TokenType.AWS],
    ['Simulator', TokenType.SIMULATOR],
    ['true', TokenType.BOOLEAN],
    ['false', TokenType.BOOLEAN],
    ['pi', TokenType.PI],
    ['theta', TokenType.THETA]
  ]);

  constructor(input: string) {
    this.input = input;
  }

  tokenize(): Token[] {
    const tokens: Token[] = [];
    
    while (this.position < this.input.length) {
      const token = this.nextToken();
      if (token.type !== TokenType.WHITESPACE && token.type !== TokenType.COMMENT) {
        tokens.push(token);
      }
    }
    
    tokens.push({
      type: TokenType.EOF,
      value: '',
      line: this.line,
      column: this.column
    });
    
    return tokens;
  }

  private nextToken(): Token {
    this.skipWhitespace();
    
    if (this.position >= this.input.length) {
      return this.makeToken(TokenType.EOF, '');
    }

    const char = this.input[this.position];
    
    // Comments
    if (char === '/' && this.peek() === '/') {
      return this.readLineComment();
    }
    
    if (char === '/' && this.peek() === '*') {
      return this.readBlockComment();
    }

    // Numbers
    if (this.isDigit(char)) {
      return this.readNumber();
    }

    // Strings
    if (char === '"') {
      return this.readString();
    }

    // Identifiers and keywords
    if (this.isAlpha(char) || char === '_') {
      return this.readIdentifier();
    }

    // Single character tokens
    switch (char) {
      case '+': return this.makeToken(TokenType.PLUS, this.advance());
      case '-': return this.makeToken(TokenType.MINUS, this.advance());
      case '*': return this.makeToken(TokenType.MULTIPLY, this.advance());
      case '/': return this.makeToken(TokenType.DIVIDE, this.advance());
      case '=': return this.makeToken(TokenType.ASSIGN, this.advance());
      case '(': return this.makeToken(TokenType.LPAREN, this.advance());
      case ')': return this.makeToken(TokenType.RPAREN, this.advance());
      case '{': return this.makeToken(TokenType.LBRACE, this.advance());
      case '}': return this.makeToken(TokenType.RBRACE, this.advance());
      case '[': return this.makeToken(TokenType.LBRACKET, this.advance());
      case ']': return this.makeToken(TokenType.RBRACKET, this.advance());
      case ',': return this.makeToken(TokenType.COMMA, this.advance());
      case ':': return this.makeToken(TokenType.COLON, this.advance());
      case ';': return this.makeToken(TokenType.SEMICOLON, this.advance());
      default:
        throw new Error(`Unexpected character: ${char} at line ${this.line}, column ${this.column}`);
    }
  }

  private readNumber(): Token {
    const start = this.position;
    let hasDecimal = false;
    
    while (this.position < this.input.length && 
           (this.isDigit(this.input[this.position]) || 
            (!hasDecimal && this.input[this.position] === '.'))) {
      if (this.input[this.position] === '.') {
        hasDecimal = true;
      }
      this.advance();
    }
    
    const value = this.input.substring(start, this.position);
    return this.makeToken(hasDecimal ? TokenType.REAL : TokenType.INTEGER, value);
  }

  private readString(): Token {
    const start = this.position;
    this.advance(); // Skip opening quote
    
    while (this.position < this.input.length && this.input[this.position] !== '"') {
      this.advance();
    }
    
    if (this.position >= this.input.length) {
      throw new Error(`Unterminated string at line ${this.line}`);
    }
    
    this.advance(); // Skip closing quote
    const value = this.input.substring(start, this.position);
    return this.makeToken(TokenType.STRING, value);
  }

  private readIdentifier(): Token {
    const start = this.position;
    
    while (this.position < this.input.length && 
           (this.isAlphaNumeric(this.input[this.position]) || this.input[this.position] === '_')) {
      this.advance();
    }
    
    const value = this.input.substring(start, this.position);
    const type = this.keywords.get(value) || TokenType.IDENTIFIER;
    return this.makeToken(type, value);
  }

  private readLineComment(): Token {
    const start = this.position;
    
    while (this.position < this.input.length && this.input[this.position] !== '\n') {
      this.advance();
    }
    
    const value = this.input.substring(start, this.position);
    return this.makeToken(TokenType.COMMENT, value);
  }

  private readBlockComment(): Token {
    const start = this.position;
    this.advance(); // Skip '/'
    this.advance(); // Skip '*'
    
    while (this.position < this.input.length - 1) {
      if (this.input[this.position] === '*' && this.input[this.position + 1] === '/') {
        this.advance(); // Skip '*'
        this.advance(); // Skip '/'
        break;
      }
      this.advance();
    }
    
    const value = this.input.substring(start, this.position);
    return this.makeToken(TokenType.COMMENT, value);
  }

  private skipWhitespace(): void {
    while (this.position < this.input.length && this.isWhitespace(this.input[this.position])) {
      this.advance();
    }
  }

  private advance(): string {
    const char = this.input[this.position];
    this.position++;
    
    if (char === '\n') {
      this.line++;
      this.column = 1;
    } else {
      this.column++;
    }
    
    return char;
  }

  private peek(offset: number = 1): string {
    const pos = this.position + offset;
    return pos < this.input.length ? this.input[pos] : '\0';
  }

  private makeToken(type: TokenType, value: string): Token {
    return {
      type,
      value,
      line: this.line,
      column: this.column - value.length
    };
  }

  private isDigit(char: string): boolean {
    return char >= '0' && char <= '9';
  }

  private isAlpha(char: string): boolean {
    return (char >= 'a' && char <= 'z') || (char >= 'A' && char <= 'Z');
  }

  private isAlphaNumeric(char: string): boolean {
    return this.isAlpha(char) || this.isDigit(char);
  }

  private isWhitespace(char: string): boolean {
    return char === ' ' || char === '\t' || char === '\r' || char === '\n';
  }
}

File: synapse-qnet/src/lexer.py
"""
Synapse QNet Lexer
Tokenizes quantum network DSL syntax
"""

import re
from enum import Enum, auto
from typing import List, NamedTuple, Optional
from dataclasses import dataclass

class TokenType(Enum):
    # Literals
    NUMBER = auto()
    STRING = auto()
    QUANTUM_STATE = auto()  # |ψ⟩, |00⟩, |Φ+⟩
    
    # Identifiers
    IDENTIFIER = auto()
    
    # Keywords
    NETWORK = auto()
    NODE = auto()
    CHANNEL = auto()
    PROTOCOL = auto()
    SIMULATION = auto()
    HARDWARE = auto()
    
    # Quantum Keywords
    PREPARE = auto()
    MEASURE = auto()
    ENTANGLE = auto()
    TELEPORT = auto()
    
    # QKD Keywords  
    BB84 = auto()
    E91 = auto()
    SIFTING = auto()
    ERROR_CORRECTION = auto()
    
    # Operators
    ARROW = auto()      # ->
    TENSOR = auto()     # ⊗
    DAGGER = auto()     # †
    
    # Delimiters
    LBRACE = auto()     # {
    RBRACE = auto()     # }
    LPAREN = auto()     # (
    RPAREN = auto()     # )
    LBRACKET = auto()   # [
    RBRACKET = auto()   # ]
    
    # Punctuation
    COLON = auto()      # :
    SEMICOLON = auto()  # ;
    COMMA = auto()      # ,
    DOT = auto()        # .
    
    # Special
    NEWLINE = auto()
    EOF = auto()
    COMMENT = auto()

@dataclass
class Token:
    type: TokenType
    value: str
    line: int
    column: int

class SynapseQLexer:
    """Lexer for Synapse QNet quantum network DSL"""
    
    KEYWORDS = {
        'network': TokenType.NETWORK,
        'node': TokenType.NODE,
        'channel': TokenType.CHANNEL,
        'protocol': TokenType.PROTOCOL,
        'simulation': TokenType.SIMULATION,
        'hardware': TokenType.HARDWARE,
        'prepare': TokenType.PREPARE,
        'measure': TokenType.MEASURE,
        'entangle': TokenType.ENTANGLE,
        'teleport': TokenType.TELEPORT,
        'BB84': TokenType.BB84,
        'E91': TokenType.E91,
        'sifting': TokenType.SIFTING,
        'error_correction': TokenType.ERROR_CORRECTION,
    }
    
    # Quantum state regex: |ψ⟩, |00⟩, |Φ+⟩, etc.
    QUANTUM_STATE_PATTERN = r'\|[^⟩]+⟩'
    
    # Number patterns
    NUMBER_PATTERN = r'-?\d+\.?\d*([eE][+-]?\d+)?'
    
    # String patterns
    STRING_PATTERN = r'"[^"]*"'
    
    # Identifier pattern
    IDENTIFIER_PATTERN = r'[a-zA-Z_][a-zA-Z0-9_]*'
    
    # Comment pattern
    COMMENT_PATTERN = r'//.*$'
    
    def __init__(self, text: str):
        self.text = text
        self.pos = 0
        self.line = 1
        self.column = 1
        self.tokens: List[Token] = []
    
    def error(self, message: str) -> None:
        raise SyntaxError(f"Line {self.line}, Column {self.column}: {message}")
    
    def peek(self, offset: int = 0) -> Optional[str]:
        """Peek at character at current position + offset"""
        pos = self.pos + offset
        if pos < len(self.text):
            return self.text[pos]
        return None
    
    def advance(self) -> Optional[str]:
        """Move to next character and return current"""
        if self.pos >= len(self.text):
            return None
            
        char = self.text[self.pos]
        self.pos += 1
        
        if char == '\n':
            self.line += 1
            self.column = 1
        else:
            self.column += 1
            
        return char
    
    def skip_whitespace(self) -> None:
        """Skip whitespace except newlines"""
        while self.peek() and self.peek() in ' \t\r':
            self.advance()
    
    def read_number(self) -> Token:
        """Read numeric literal"""
        start_column = self.column
        value = ''
        
        # Handle negative numbers
        if self.peek() == '-':
            value += self.advance()
        
        # Read digits and decimal point
        while self.peek() and (self.peek().isdigit() or self.peek() in '.eE+-'):
            value += self.advance()
        
        return Token(TokenType.NUMBER, value, self.line, start_column)
    
    def read_string(self) -> Token:
        """Read string literal"""
        start_column = self.column
        self.advance()  # Skip opening quote
        value = ''
        
        while self.peek() and self.peek() != '"':
            if self.peek() == '\\':
                self.advance()  # Skip escape character
                escaped = self.advance()
                if escaped == 'n':
                    value += '\n'
                elif escaped == 't':
                    value += '\t'
                elif escaped == '\\':
                    value += '\\'
                elif escaped == '"':
                    value += '"'
                else:
                    value += escaped
            else:
                value += self.advance()
        
        if not self.peek():
            self.error("Unterminated string literal")
        
        self.advance()  # Skip closing quote
        return Token(TokenType.STRING, value, self.line, start_column)
    
    def read_quantum_state(self) -> Token:
        """Read quantum state notation like |ψ⟩"""
        start_column = self.column
        value = ''
        
        # Read until closing ⟩
        while self.peek() and self.peek() != '⟩':
            value += self.advance()
        
        if not self.peek():
            self.error("Unterminated quantum state")
        
        value += self.advance()  # Include closing ⟩
        return Token(TokenType.QUANTUM_STATE, value, self.line, start_column)
    
    def read_identifier(self) -> Token:
        """Read identifier or keyword"""
        start_column = self.column
        value = ''
        
        while (self.peek() and 
               (self.peek().isalnum() or self.peek() in '_')):
            value += self.advance()
        
        # Check if it's a keyword
        token_type = self.KEYWORDS.get(value, TokenType.IDENTIFIER)
        return Token(token_type, value, self.line, start_column)
    
    def read_comment(self) -> Token:
        """Read single-line comment"""
        start_column = self.column
        value = ''
        
        while self.peek() and self.peek() != '\n':
            value += self.advance()
        
        return Token(TokenType.COMMENT, value, self.line, start_column)
    
    def tokenize(self) -> List[Token]:
        """Tokenize the entire input"""
        while self.pos < len(self.text):
            self.skip_whitespace()
            
            if not self.peek():
                break
            
            char = self.peek()
            
            # Newlines
            if char == '\n':
                token = Token(TokenType.NEWLINE, char, self.line, self.column)
                self.tokens.append(token)
                self.advance()
            
            # Comments
            elif char == '/' and self.peek(1) == '/':
                token = self.read_comment()
                self.tokens.append(token)
            
            # Numbers
            elif char.isdigit() or (char == '-' and self.peek(1) and self.peek(1).isdigit()):
                token = self.read_number()
                self.tokens.append(token)
            
            # Strings
            elif char == '"':
                token = self.read_string()
                self.tokens.append(token)
            
            # Quantum states
            elif char == '|':
                token = self.read_quantum_state()
                self.tokens.append(token)
            
            # Identifiers and keywords
            elif char.isalpha() or char == '_':
                token = self.read_identifier()
                self.tokens.append(token)
            
            # Operators and punctuation
            elif char == '-' and self.peek(1) == '>':
                token = Token(TokenType.ARROW, '->', self.line, self.column)
                self.tokens.append(token)
                self.advance()
                self.advance()
            
            elif char == '⊗':
                token = Token(TokenType.TENSOR, char, self.line, self.column)
                self.tokens.append(token)
                self.advance()
            
            elif char == '†':
                token = Token(TokenType.DAGGER, char, self.line, self.column)
                self.tokens.append(token)
                self.advance()
            
            elif char == '{':
                token = Token(TokenType.LBRACE, char, self.line, self.column)
                self.tokens.append(token)
                self.advance()
            
            elif char == '}':
                token = Token(TokenType.RBRACE, char, self.line, self.column)
                self.tokens.append(token)
                self.advance()
            
            elif char == '(':
                token = Token(TokenType.LPAREN, char, self.line, self.column)
                self.tokens.append(token)
                self.advance()
            
            elif char == ')':
                token = Token(TokenType.RPAREN, char, self.line, self.column)
                self.tokens.append(token)
                self.advance()
            
            elif char == '[':
                token = Token(TokenType.LBRACKET, char, self.line, self.column)
                self.tokens.append(token)
                self.advance()
            
            elif char == ']':
                token = Token(TokenType.RBRACKET, char, self.line, self.column)
                self.tokens.append(token)
                self.advance()
            
            elif char == ':':
                token = Token(TokenType.COLON, char, self.line, self.column)
                self.tokens.append(token)
                self.advance()
            
            elif char == ';':
                token = Token(TokenType.SEMICOLON, char, self.line, self.column)
                self.tokens.append(token)
                self.advance()
            
            elif char == ',':
                token = Token(TokenType.COMMA, char, self.line, self.column)
                self.tokens.append(token)
                self.advance()
            
            elif char == '.':
                token = Token(TokenType.DOT, char, self.line, self.column)
                self.tokens.append(token)
                self.advance()
            
            else:
                self.error(f"Unexpected character: {char}")
        
        # Add EOF token
        self.tokens.append(Token(TokenType.EOF, '', self.line, self.column))
        return self.tokens

def main():
    """Test the lexer with sample code"""
    sample_code = '''
    network TestNet {
        node Alice {
            type: "endpoint"
            location: [40.7128, -74.0060]
        }
        
        channel AliceBob {
            connect: Alice -> Bob
            medium: "fiber"
            loss_db_per_km: 0.2
        }
    }
    
    protocol BB84 {
        prepare |ψ⟩ in basis(random[H,V,D,A])
        measure in basis(random[H,V,D,A])
    }
    '''
    
    lexer = SynapseQLexer(sample_code)
    tokens = lexer.tokenize()
    
    for token in tokens:
        if token.type != TokenType.NEWLINE:  # Skip newlines for cleaner output
            print(f"{token.type.name:15} | {token.value:20} | Line {token.line}, Col {token.column}")

if __name__ == "__main__":
    main()